### PR TITLE
Master merge color picker palette fja

### PR DESCRIPTION
**Pull request by fja-odoo**
--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by fja-odoo** [comment]
@qsm-odoo Cherry picked the 2 commits from https://github.com/odoo/odoo/pull/45872 into the PR.

File: addons/web/static/src/js/fields/basic_fields.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ var utils = require('web.utils');
 var view_dialogs = require('web.view_dialogs');
 var field_utils = require('web.field_utils');
 var time = require('web.time');
-var ColorpickerDialog = require('web.ColorpickerDialog');
+var ColorpickerDialog = require('web.Colorpicker').ColorpickerDialog;
```
Comment (qsm-odoo) [inline addons/web/static/src/js/fields/basic_fields.js:26]: For this, I like `const { ColorpickerDialog } = require('web.Colorpicker');`. This avoids importing it under another name making grepping the code less easy.

File: addons/web/static/src/js/widgets/colorpicker.js [hunk 2]
```diff
@@ -1,22 +1,22 @@
-odoo.define('web.ColorpickerDialog', function (require) {
+odoo.define('web.Colorpicker', function (require) {
```
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/colorpicker.js:1]: Don't force me to repeat that again please ;) You move/rename a file? You do that in a dedicated commit.

File: addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js [hunk 3]
```diff
@@ -3,7 +3,7 @@ odoo.define('web_editor.ColorPalette', function (require) {
 
 const core = require('web.core');
 const session = require('web.session');
-const ColorpickerDialog = require('web.ColorpickerDialog');
+const {ColorpickerDialog, ColorpickerWidget} = require('web.Colorpicker');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js]: You actually use it here :p Use it everywhere :p

File: addons/web/static/src/js/widgets/colorpicker.js [hunk 4]
```diff
@@ -40,17 +40,19 @@ var ColorpickerWidget = Widget.extend({
...
-            self._onMouseMovePicker(ev);
-            self._onMouseMoveSlider(ev);
-            self._onMouseMoveOpacitySlider(ev);
-        }, 10));
-        $body.on('mouseup.colorpicker', _.throttle(function (ev) {
-            self.pickerFlag = false;
-            self.sliderFlag = false;
-            self.opacitySliderFlag = false;
+        const $document = $(document);
+        $document.on('mousemove.colorpicker', _.throttle((ev) => {
```
Comment (qsm-odoo) [inline addons/web/static/src/js/widgets/colorpicker.js:44]: Document instead of body?

File: addons/web/static/src/xml/colorpicker.xml [hunk 5]
```diff
@@ -1,46 +1,32 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <templates id="template" xml:space="preserve">
 
-<t t-name="ColorpickerDialog">
+<t t-name="Colorpicker">
```
Comment (qsm-odoo) [inline addons/web/static/src/xml/colorpicker.xml:4]: Is this in the right commit?

File: addons/web_editor/static/src/js/editor/rte.summernote.js [hunk 6]
```diff
@@ -30,11 +30,15 @@ var tplButton = renderer.getTemplate().button;
 var tplIconButton = renderer.getTemplate().iconButton;
 var tplDropdown = renderer.getTemplate().dropdown;
 
-var applyColor = function (target, eventName, color) {
+const processAndApplyColor = function (target, eventName, color) {
+    if (!color) {
+        color = 'inherit';
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/rte.summernote.js:35]: Why forcing inherit? (was not the case before, right?)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 7]
```diff
@@ -1021,9 +1034,13 @@ const ColorpickerUserValueWidget = SelectUserValueWidget.extend({
      * @override
      */
     getValue: function (methodName) {
+        // previewColor takes priority over customColorValue !
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1037]: No need of that comment, it's explicit enough ;)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 8]
```diff
@@ -1061,7 +1078,7 @@ const ColorpickerUserValueWidget = SelectUserValueWidget.extend({
         const oldColorPalette = this.colorPalette;
         this.colorPalette = new ColorPaletteWidget(this, options);
         if (oldColorPalette) {
-            return this.colorPalette.insertAfter(oldColorPalette.el).then(() => {
+            return this.colorPalette.replace(oldColorPalette.el).then(() => {
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1081]: Why? Kinda prefer to let the widget destroy itself entirely by itself.

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 9]
```diff
@@ -1093,6 +1110,16 @@ const ColorpickerUserValueWidget = SelectUserValueWidget.extend({
     // Handlers
     //--------------------------------------------------------------------------
 
+    /**
+     * Called when a custom color is selected -> preview the color
+     * and set the current value. Update of this value on close !
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1115]: You forgot to add an emoji after your exclamation mark ;)

File: addons/web_editor/static/src/js/editor/snippets.options.js [hunk 10]
```diff
@@ -1125,14 +1153,15 @@ const ColorpickerUserValueWidget = SelectUserValueWidget.extend({
...
      */
-    _onColorReset: function (ev) {
-        this._value = '';
-        this._onUserValueChange(ev);
+    _onClick: function (ev) {
+        // Do not close the colorpalette on colorpicker click
+        if (ev.target.closest('.o_colorpicker_widget')) {
+            ev.stopPropagation();
+            return;
+        }
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/editor/snippets.options.js:1163]: Mmh. Still not sure what I prefer to handle this :) I kinda like flagging the event in the widget itself and checking the flag here. It has the advantage to:
- not force stopPropagation in the widget itself
- allow the parent to stopPropagation of events of internal widgets without knowing their internal structure/main class

But this is fine by me too. I let you judge.

File: addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js [hunk 11]
```diff
@@ -3,7 +3,7 @@ odoo.define('web_editor.ColorPalette', function (require) {
 
 const core = require('web.core');
 const session = require('web.session');
-const {ColorpickerDialog, ColorpickerWidget} = require('web.Colorpicker');
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js]: Just had to remove ColorpickerDialog ;)

File: addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js [hunk 12]
```diff
@@ -71,7 +73,7 @@ const ColorPaletteWidget = Widget.extend({
      * @override
      */
     start: function () {
-        const res = this._super.apply(this, arguments);
+        const proms = [this._super.apply(this, arguments)];
```
Comment (qsm-odoo) [inline addons/web_editor/static/src/js/wysiwyg/widgets/color_palette.js:76]: Feels like you can just await without any slowdown here. Instead of building a proms array then waiting for all

**Comment by fja-odoo** [comment]
@qsm-odoo I Will need this fix to satisfy SBU request. https://github.com/odoo/odoo/pull/47189
Should I cherry pick or we merge and fp soon ?

**Comment by qsm-odoo** [comment]
I'll merge it before lunch ;)

**Comment by qsm-odoo** [comment]
You can c/c on the beginning on your branch meanwhile of course

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 10 above)
Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: I think you are right. When you say flagging the event you mean passing additional data to it in the colorpickerWidget ? Is it possible without customEvent ?

File: addons/web_editor/static/src/js/editor/rte.summernote.js (hunk 6 above)
Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/rte.summernote.js:35]: it was the case in color reset. Now when the color picked is an empty string it is similar to a color reset.

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 10 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: An event is just an object, you can just add a property to it.

Widget:
```js
myHandler: function (ev) {
    ev.__thisEventWasSeen = true;
},
```
Parent:
```js
if (ev.__thisEventWasSeen) {
    ev.stopPropagation();
}
```
I would name the flag with the name of the widget class or something like that

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: (I thought about making that generic to all widgets at some point: add an automatic handler to all widget which flags any event with the name of the widget... but the case is not that common anyway so too crappy to have code running for that).

Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: I know that but it looks like you can not alter an event that is bubbling. It looks like it is recreated each time it bubble. If it is captured on the same element no problem but on a parent the modifications are not replicated.

Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: Looks like it is working inside of the same widget but not from a wiget to another

Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: You are probably using the jQuery event object instead of the JavaScript event

Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: I have to do this then

Widget:

```js
myHandler: function (ev) {
    ev.originalEvent.__thisEventWasSeen = true;
},
```
Parent:

```js
if (ev.originalEvent.__thisEventWasSeen) {
    ev.stopPropagation();
}
```

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 8 above)
Comment (fja-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1081]: At the beginning of the dev the widget was instantiated while the modal was open which caused the width to shrink due to the scrollbar appearing (2 color_palette is longer than max-width). That was displaying the cursor at the wrong place in the colorpicker.
This is no longer an issue so I will revert that.

File: addons/web/static/src/js/widgets/colorpicker.js (hunk 4 above)
Comment (fja-odoo) [reply addons/web/static/src/js/widgets/colorpicker.js:44]: I would love to tell you why but I have not looked into it. When moving it from the modal to the left panel it no longer worked. document did the job and I don't see an issue with it so I went with that.

File: addons/web_editor/static/src/js/editor/snippets.options.js (hunk 10 above)
Comment (qsm-odoo) [reply addons/web_editor/static/src/js/editor/snippets.options.js:1163]: :+1:

File: addons/web/static/src/js/widgets/colorpicker.js (hunk 4 above)
Comment (qsm-odoo) [reply addons/web/static/src/js/widgets/colorpicker.js:44]: No issue indeed, I was just curious as I don't know the difference. Maybe add a comment so that nobody (including me) changes it

**Comment by qsm-odoo** [comment]
FJA's notes about the implementation:
```
Tough to make this work. Had to rethink part of the colorpalette
behaviours.
The colorpicker used to close everytime we clicked on it, this is no
longer the case as we do not want it to close each time we try a new
custom color in the colorpicker.
This introduced another issue in the fact that we can not call a value
change each time the colorpicker has a new value because that will
reload the color_palette resulting in lots of useless rerender and
rounding of the values on each rerender. Making the colorpicker look
cluncky...
We have to update the value of our widget on close but only when the
value has changed else there will be useless updates on each rerender
causing issues in other options.
This all leads us to add _customColorValue which has a behaviour a
little similar to _previewColor. It will hold the last customColor
selected and replace the widget value on close. hovering a palette
color will take priority over our customValue for the preview.

The color_palette is now a button like the other colors, no need for a
color_reset event anymore.
It is removed and replaced with a color_picked event with color
being an empty string. The clear button now also work on hover.

The header colorpalette for text and text background color now has the
exact same behaviour as the background colorpalette: adding color
preview on hover.

The color_palette now also close on enter key press.
```

File: addons/web/static/src/xml/colorpicker.xml [hunk 13]
```diff
@@ -3,44 +3,30 @@
...
+            <div t-if="widget.options.colorPreview" class="o_color_preview_parent">
+                <span class="rounded-0 o_color_preview"/>
+            </div>
+            <div class="o_hex_div"><input type="text" class="o_hex_input" data-color-method="hex" size="1"/>hex</div>
+            <div class="o_rgba_div">
+                <input type="text" class="o_red_input" data-color-method="rgb" size="1"/>
+                <input type="text" class="o_green_input" data-color-method="rgb" size="1"/>
+                <input type="text" class="o_blue_input" data-color-method="rgb" size="1"/>
+                <t t-if="!widget.options.noTransparency">
+                    <input t-if="!widget.options.noTransparency" type="text" class="o_opacity_input" data-color-method="opacity" size="1"/>
```
Comment (qsm-odoo) [inline addons/web/static/src/xml/colorpicker.xml:25]: To be sure?

**Comment by qsm-odoo** [comment]
@robodoo rebase-ff r+

**Comment by qsm-odoo** [comment]
@robodoo r-

**Comment by qsm-odoo** [comment]
@robodoo rebase-ff r+
